mir_perception/mir_handle_detection/ros/src/drawer_handle_perceiver.cpp

```diff
@@ -0,0 +1,256 @@
+    this->seg.setModelType(pcl::SACMODEL_PLANE);
+    this->seg.setMethodType(pcl::SAC_RANSAC);
+    this->seg.setDistanceThreshold(seg_dist_threshold);
+
+    this->project_inliers.setModelType(pcl::SACMODEL_NORMAL_PARALLEL_PLANE);
+    this->project_inliers.setCopyAllData(false);
+
+    float polygon_prism_height_min, polygon_prism_height_max;
+    nh.param<float>("polygon_prism_height_min", polygon_prism_height_min, 0.01);
+    nh.param<float>("polygon_prism_height_max", polygon_prism_height_max, 0.1);
+    this->extract_polygonal_prism.setHeightLimits(polygon_prism_height_min, polygon_prism_height_max);
+
+    this->extract_indices.setNegative(false);
+
+    float cluster_tolerance;
+    int min_cluster_size, max_cluster_size;
+    nh.param<float>("cluster_tolerance", cluster_tolerance, 0.02);
+    nh.param<int>("min_cluster_size", min_cluster_size, 50);
+    nh.param<int>("max_cluster_size", max_cluster_size, 10000);
+    this->euclidean_cluster_extraction.setClusterTolerance(cluster_tolerance);
+    this->euclidean_cluster_extraction.setMinClusterSize(min_cluster_size);
+    this->euclidean_cluster_extraction.setMaxClusterSize(max_cluster_size);
+
+    this->listening = false;
+}
+
+DrawerHandlePerceiver::~DrawerHandlePerceiver()
+{
+}
+
+void DrawerHandlePerceiver::eventInCallback(const std_msgs::String::ConstPtr &msg)
+{
+    if (msg->data == "e_start")
+    {
+        ROS_INFO_STREAM("starting listening");
+        this->listening = true;
+    }
+    else if (msg->data == "e_stop")
+    {
+        ROS_INFO_STREAM("stoping listening");
+        this->listening = false;
+    }
+}
+
+void DrawerHandlePerceiver::pcCallback(const sensor_msgs::PointCloud2::ConstPtr &msg)
+{
+    if (!this->listening)
+        return;
+
+    sensor_msgs::PointCloud2 msg_transformed;
+    bool success = this->transformPC(msg, msg_transformed);
+    if (!success)
+        return;
+
+    PCloudT::Ptr pc_input(new PCloudT);
+    pcl::fromROSMsg(msg_transformed, *pc_input);
+
+    PCloudT::Ptr pc_downsampled(new PCloudT);
+    this->downSamplePC(pc_input, pc_downsampled, 3);
+
+    PCloudT::Ptr pc_passthrough_filtered(new PCloudT);
+    this->passthroughFilterPC(pc_downsampled, pc_passthrough_filtered);
+
+    PCloudT::Ptr pc_filtered(new PCloudT);
+    this->voxel_grid_filter.setInputCloud(pc_passthrough_filtered);
+    this->voxel_grid_filter.filter(*pc_filtered);
+
+    PCloudT::Ptr pc_segmented(new PCloudT);
+    this->extractPlaneOutlier(pc_filtered, pc_passthrough_filtered, pc_segmented);
+
+    Eigen::Vector4f closest_centroid(0.0, 0.0, 0.0, 0.0);
+    bool cluster_success = this->getClosestCluster(pc_segmented, closest_centroid);
+    if (!cluster_success)
+        return;
+
+    geometry_msgs::PoseStamped pose_stamped;
+    pose_stamped.pose.position.x = closest_centroid[0];
+    pose_stamped.pose.position.y = closest_centroid[1];
+    pose_stamped.pose.position.z = closest_centroid[2];
+    pose_stamped.pose.orientation.w = 1.0;
```

Comment:
create an issue for this (i.e. actually calculate the orientation of the cluster here instead of hardcoding it)

Reply:
https://github.com/b-it-bots/mas_industrial_robotics/issues/102